File: src/shared/kartEngine/KartMove/GoKart.ts
/**
 * GoKart class的TypeScript实现
 * 与Lua版本完全一致
 */

import { UnityVector3 } from "../KartShared/UnityMath";
import { BoostKind } from "./BoostKind";
import { GameObject } from "../KartShared/RobloxUnityAdapter";

export class GoKart {
    // 私有属性
    private stuck_: boolean = false;
    private valid_: boolean = true;
    private forcing_: boolean = false;
    private isInResetState_: boolean = false;
    private status_: number = 0;
    
    // 公共成员
    public m_kart: unknown = undefined; // Roblox对象
    public controller_: unknown = undefined;
    public m_KartWLVel: UnityVector3 = UnityVector3.zero;
    public m_KartLAVel: UnityVector3 = UnityVector3.zero;
    public m_KartRealVelocity: UnityVector3 = UnityVector3.zero;

    constructor() {
        // 属性初始化在声明时完成
    }

    // 属性访问器
    public GetStuck(): boolean {
        return this.stuck_;
    }

    public SetStuck(value: boolean): void {
        this.stuck_ = value;
    }

    public GetValid(): boolean {
        return this.valid_;
    }

    public SetValid(value: boolean): void {
        if (this.valid_ !== value) {
            this.valid_ = value;
        }
    }

    public GetForcing(): boolean {
        return this.forcing_;
    }

    public SetForcing(value: boolean): void {
        this.forcing_ = value;
    }

    public GetIsInResetState(): boolean {
        return this.isInResetState_;
    }

    public SetIsInResetState(value: boolean): void {
        this.isInResetState_ = value;
    }

    // 虚函数 - 子类可以重写
    public basicAction(tick: number): void {
        // 基类的空实现
    }

    public setReKart(controller: unknown, wheels: unknown): void {
        this.controller_ = controller;
        // 严格按照Lua代码：if controller and controller.gameObject then
        if (controller && typeOf(controller) === "table") {
            const ctrl = controller as Record<string, unknown>;
            if (ctrl.gameObject) {
                this.m_kart = ctrl.gameObject as GameObject;
            } else {
                warn("GoKart: 无法获取Roblox对象");
            }
        } else {
            warn("GoKart: 无法获取Roblox对象");
        }
    }

    public Warp(pos: any, rot: unknown, flush?: boolean, resetVel?: boolean): void {
        if (this.m_kart && (this.m_kart as Record<string, unknown>).transform) {
            const kart = this.m_kart as Record<string, Record<string, unknown>>;
            const transform = kart.transform as unknown as {
                setLocalPosition: (pos: Vector3) => void;
                setLocalRotation: (rot: unknown) => void;
            };
            if (transform) {
                transform.setLocalPosition(pos);
                transform.setLocalRotation(rot);
            }
        }
        
        if (flush) {
            // 刷新逻辑
        }
        
        if (resetVel) {
            this.m_KartWLVel = UnityVector3.zero;
            this.m_KartLAVel = UnityVector3.zero;
        }
    }

    public isRealBoost(): boolean {
        return false;
    }

    public isRealBoostWithKind(kind: BoostKind): boolean {
        return false;
    }

    public isItemBoost(): boolean {
        return false;
    }

    public isItemBoostWithKind(kind: BoostKind): boolean {
        return false;
    }

    public isZoneBoost(): boolean {
        return false;
    }

    public isZoneBoostWithKind(kind: BoostKind): boolean {
        return false;
    }

    public isBoost(kind: BoostKind): boolean {
        return false;
    }

    public ResetForRestarting(): void {
        this.m_KartWLVel = new UnityVector3(0, 0, 0);
        this.m_KartLAVel = new UnityVector3(0, 0, 0);
        this.m_KartRealVelocity = new UnityVector3(0, 0, 0);
        this.stuck_ = false;
        this.valid_ = true;
        this.forcing_ = false;
        this.status_ = 0;
    }
}

export default GoKart;

File: sources/roblox-lua-items-network/shared/kartEngine/KartMove/GoKart.lua
-- GoKart class的Lua等效实现
local UnityMath = require(script.Parent.Parent.KartShared.UnityMath)
local Vector3 = UnityMath.Vector3
local BoostKind = require(script.Parent.BoostKind)

local GoKart = {}
GoKart.__index = GoKart

function GoKart.new()
    local self = setmetatable({}, GoKart)
    
    -- 属性初始化
    self.stuck_ = false
    self.valid_ = true
    self.forcing_ = false
    self.isInResetState_ = false
    self.status_ = 0
    
    -- 公共成员
    self.m_kart = nil
    self.controller_ = nil
    self.m_KartWLVel = Vector3.zero
    self.m_KartLAVel = Vector3.zero
    self.m_KartRealVelocity = Vector3.zero
    
    return self
end

-- 属性访问器
function GoKart:GetStuck()
    return self.stuck_
end

function GoKart:SetStuck(value)
    self.stuck_ = value
end

function GoKart:GetValid()
    return self.valid_
end

function GoKart:SetValid(value)
    if self.valid_ ~= value then
        self.valid_ = value
    end
end

function GoKart:GetForcing()
    return self.forcing_
end

function GoKart:SetForcing(value)
    self.forcing_ = value
end

function GoKart:GetIsInResetState()
    return self.isInResetState_
end

function GoKart:SetIsInResetState(value)
    self.isInResetState_ = value
end

-- 虚函数 - 子类可以重写
function GoKart:basicAction(tick)
    -- 基类的空实现
end

function GoKart:setReKart(controller, wheels)
    self.controller_ = controller
    -- 获取真正的Roblox对象
    if controller and controller.gameObject then
        self.m_kart = controller.gameObject
    else
        warn("GoKart: 无法获取Roblox对象")
    end
end

function GoKart:Warp(pos, rot, flush, resetVel)
    if self.m_kart and self.m_kart.transform then
        self.m_kart.transform.localPosition = pos
        self.m_kart.transform.localRotation = rot
    end
    
    if flush then
        -- 刷新逻辑
    end
    
    if resetVel then
        self.m_KartWLVel = Vector3.zero
        self.m_KartLAVel = Vector3.zero
    end
end

function GoKart:isRealBoost()
    return false
end

function GoKart:isRealBoostWithKind(kind)
    return false
end

function GoKart:isItemBoost()
    return false
end

function GoKart:isItemBoostWithKind(kind)
    return false
end

function GoKart:isZoneBoost()
    return false
end

function GoKart:isZoneBoostWithKind(kind)
    return false
end

function GoKart:isBoost(kind)
    return false
end

function GoKart:ResetForRestarting()
    self.m_KartWLVel = Vector3.zero
    self.m_KartLAVel = Vector3.zero
    self.m_KartRealVelocity = Vector3.zero
    self.stuck_ = false
    self.valid_ = true
    self.forcing_ = false
    self.status_ = 0
end

return GoKart